libdnf/repo/Repo.cpp

```diff
@@ -846,27 +846,34 @@ std::vector<Key> Repo::Impl::retrieve(const std::string & url)
     return keyInfos;
 }
 
-static void killGpgAgent(gpgme_ctx_t context, const std::string & gpgDir)
-{
-    auto logger(Log::getLogger());
-
-    auto gpgErr = gpgme_set_protocol(context, GPGME_PROTOCOL_ASSUAN);
-    if (gpgErr != GPG_ERR_NO_ERROR) {
-        auto msg = tfm::format(_("%s: gpgme_set_protocol(): %s"), __func__, gpgme_strerror(gpgErr));
-        logger->warning(msg);
-        return;
-    }
-    std::string gpgAgentSock = gpgDir + "/S.gpg-agent";
-    gpgErr = gpgme_ctx_set_engine_info(context, GPGME_PROTOCOL_ASSUAN, gpgAgentSock.c_str(), gpgDir.c_str());
-    if (gpgErr != GPG_ERR_NO_ERROR) {
-        auto msg = tfm::format(_("%s: gpgme_ctx_set_engine_info(): %s"), __func__, gpgme_strerror(gpgErr));
-        logger->warning(msg);
-        return;
-    }
-    gpgErr = gpgme_op_assuan_transact_ext(context, "KILLAGENT", NULL, NULL, NULL, NULL, NULL, NULL, NULL);
-    if (gpgErr != GPG_ERR_NO_ERROR) {
-        auto msg = tfm::format(_("%s: gpgme_op_assuan_transact_ext(): %s"), __func__, gpgme_strerror(gpgErr));
-        logger->debug(msg);
+/*
+ * Creates the '/run/user/$UID' directory if it doesn't exist. If this
+ * directory exists, gpgagent will create its sockets under
+ * '/run/user/$UID/gnupg'.
+ *
+ * If this directory doesn't exist, gpgagent will create its sockets in gpg
+ * home directory, which is under '/var/cache/yum/metadata/' and this was
+ * causing trouble with container images, see [1].
+ *
+ * Previous solution was to send the agent a "KILLAGENT" message, but that
+ * would cause a race condition with calling gpgme_release(), see [2], [3],
+ * [4].
+ *
+ * Since the agent doesn't clean up its sockets properly, by creating this
+ * directory we make sure they are in a place that is not causing trouble with
+ * container images.
+ *
+ * [1] https://bugzilla.redhat.com/show_bug.cgi?id=1650266
+ * [2] https://bugzilla.redhat.com/show_bug.cgi?id=1769831
+ * [3] https://github.com/rpm-software-management/microdnf/issues/50
+ * [4] https://bugzilla.redhat.com/show_bug.cgi?id=1781601
+ */
+static void ensure_socket_dir_exists() {
+    char dirname[32];
+    snprintf(dirname, sizeof(dirname), "/run/user/%u", getuid());
+    int res = mkdir(dirname, 0700);
+    if (res != 0 && errno != EEXIST) {
+        g_debug("Failed to create \"%s\": %d - %s\n", dirname, errno, strerror(errno));
```

Comment:
C++ code for formatting the string could be streamlined a bit, but it's fine the way it is...

However, you should use `logger.debug` instead of `g_debug` here...

Reply:
True. I'll fix it.